### PR TITLE
tab completion: fix some isDeprecated false-positives

### LINKED ISSUE #12043: Scala 2.13: REPL gives a wrong completion hint of deprecation for `GrowableBuilder`
## reproduction steps

In Scala 2.13.2, REPL

```scala
Welcome to Scala 2.13.2 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_181).
Type in expressions for evaluation. Or try :help.

scala> import scala.collection.mutable.Growable
Growable                       GrowableBuilder (deprecated)

```

## problem
The REPL show that the function is deprecated, however, there is not any deprecated annotation in the source code.
`library/scala/collection/mutable/GrowableBuilder.scala`

```scala
package scala
package collection.mutable


/** The canonical builder for collections that are growable, i.e. that support an
  * efficient `+=` method which adds an element to the collection.
  *
  * GrowableBuilders can produce only a single instance of the collection they are growing.
  *
  * @define Coll `GrowingBuilder`
  * @define coll growing builder
  */
class GrowableBuilder[Elem, To <: Growable[Elem]](protected val elems: To)
  extends Builder[Elem, To] {
```

I tried to print message to debug. 

It seems that something went wrong during check function
`repl/scala/tools/nsc/interpreter/PresentationCompilation.scala`

```scala
def isMemberDeprecated(m: Member) = m match {
  case tm: TypeMember if tm.viaView.isDeprecated || tm.viaView != NoSymbol && tm.viaView.owner.isDeprecated || !m.sym.exists =>
    true
  case _ =>
    m.sym.isDeprecated ||
    m.sym.hasGetter && m.sym.getterIn(m.sym.owner).isDeprecated
}
```

`m.sym.exists` return `false` here, so `GrowableBuilder` class is judged as deprecated.

What's more

`reflect/scala/reflect/internal/Symbols.scala`

```scala
def exists: Boolean = !isTopLevel || {
  val isSourceLoader = rawInfo match {
    case sl: SymLoader => sl.fromSource
    case _             => false
  }
  def warnIfSourceLoader(): Unit = {
    if (isSourceLoader)
      // Predef is completed early due to its autoimport; we used to get here when type checking its
      // parent LowPriorityImplicits. See comment in c5441dc for more elaboration.
      // Since the fix for scala/bug#7335 Predef parents must be defined in Predef.scala, and we should not
      // get here anymore.
      devWarning(s"calling Symbol#exists with sourcefile based symbol loader may give incorrect results.")
  }

  rawInfo load this
  rawInfo != NoType || { warnIfSourceLoader(); false }
}
```

`rawInfo` here equals `NoType`

One more step, in the same file

```scala
def rawInfo: Type = {
  // OPT: hoisting the outer reference reduces the bytecode size of this method a little which makes it more
  //      likely to inline into hot callers of .info
  val outer = Symbols.this

  var infos = this.infos
  outer.assert(infos != null, "infos must not be null")

  if (_validTo != NoPeriod) {
    val curPeriod = outer.currentPeriod
    val curPid = outer.phaseId(curPeriod)

    // skip any infos that concern later phases
    while (curPid < outer.phaseId(infos.validFrom) && infos.prev != null)
      infos = infos.prev

    if (_validTo < curPeriod) {
      // adapt any infos that come from previous runs
      val curPhase = outer.phase
      try {
        if (infos != null && outer.runId(infos.validFrom) != outer.currentRunId) {
          // scala/bug#8871 Discard all but the first element of type history. Specialization only works in the resident
          // compiler / REPL if re-run its info transformer in this run to correctly populate its
          // per-run caches, e.g. typeEnv
          adaptInfo(infos.oldest)
          infos = this.infos
        }

        //assert(runId(validTo) == currentRunId, name)
        //assert(runId(infos.validFrom) == currentRunId, name)

        if (_validTo < curPeriod) {
          infos = transformInfos(infos, curPhase, curPeriod)
        }
      } finally {
        outer.phase = curPhase
      }
    }
  }
  infos.info
}
```

So it's clearly that the wrong hint about deprecation is relative to the value of `infos.info`

I tried to print message inside `rawInfo` function, but it will lead to stackoverflow exception.

I tried to use `assert` but it seems not works only in `GrowableBuilder` and trigger error by others.

### PR DESCRIPTION
**Pull request by SethTisue**
fixes scala/bug#12043

### CODE REVIEW

**Comment by SethTisue** [comment]
as I said on the scala/bug ticket, I cannot fathom why I ever included `|| !m.sym.exists` in the first place

**Review by som-snytt — APPROVED** [review]
I had looked at this and had the same question, and also didn't have an answer. I got as far as adding some debug.

I won't ask you to use assertEquals. Membership has its privileges.

File: test/junit/scala/tools/nsc/interpreter/CompletionTest.scala [hunk 1]
```diff
@@ -206,6 +204,28 @@ class CompletionTest {
...
+
+  @Test
+  def isNotDeprecated(): Unit = {
+    val completer = setup()
+    // neither Growable nor GrowableBuilder are deprecated -- scala/bug#12043
+    val candidates = completer.complete("import scala.collection.mutable.Growable").candidates
+    assert(candidates.size == 2)
+    assert(candidates.forall(!_.isDeprecated))
+  }
+
```
Comment (retronym) [inline test/junit/scala/tools/nsc/interpreter/CompletionTest.scala:229]: I'd prefer if you define test classes variations of `@deprecated` separate from the standard library.  I think they could just be defined in this file; if not they could be defined in a previous line of the REPL session.

Comment (som-snytt) [reply test/junit/scala/tools/nsc/interpreter/CompletionTest.scala:229]: Indeed, in 5 years, 3.1 will remove `Traversable`, and in 10 years, 4.0 will rewrite collections entirely, no doubt deprecating `GrowableBuilder`!

**Review by lrytz — APPROVED** [review]
This fix LGTM (I agree with Jason's remark about the test), but we should investigate the completion results that are coming in, they look suspicious. I'll file another ticket.

**Comment by som-snytt** [comment]
Hope I did that right. Nothing good happens after midnight and before morning coffee.

**Comment by lrytz** [comment]
> I'll file another ticket

Did a PR in the end: https://github.com/scala/scala/pull/9079

**Comment by lrytz** [comment]
This PR can go in as-is, mine only brings additional cleanups.

**Pull request merged**